### PR TITLE
Envoy-gloo bump to support headerBodyTransform improvements

### PR DESCRIPTION
**Pull request by ben-taussig-solo**
# Description
 - Update envoy-gloo pulling in changes that add `queryStringParameters`, `multiValueQueryStringParameters`, and `multiValueHeaders` fields to headerBodyTransform output when `add_request_metadata` is enabled.
 - This also adds these fields to the AWS Lambda payload when the `requestTransform` setting is enabled
 - Documentation updates describing these additions are included with the PR

# Notes
 - We must update the envoy-gloo image to a released version before merging this PR
   - Additionally, build-bot will fail until this image is updated to a released version
   - The envoy-gloo image must be built from the release-v1.24 branch of envoy-gloo, not main.
BOT NOTES: 
resolves https://github.com/solo-io/solo-projects/issues/4571

### CODE REVIEW

**Comment by jbohanon** [comment]
>   * The envoy-gloo image must be built from the release-v1.24 branch of envoy-gloo, not main.

What is causing this restriction? Are we trying to not bump to envoy-gloo 1.25 in this release?

File: test/e2e/aws_test.go [hunk 1]
```diff
@@ -93,7 +93,16 @@ var _ = Describe("AWS Lambda", func() {
...
-			res, err := http.Post(fmt.Sprintf("http://%s:%d/1?param_a=value_1&param_b=value_b", "localhost", envoyPort), "application/octet-stream", &buf)
+			// form request with multi-value query params
+			req, err := http.NewRequest("POST", fmt.Sprintf("http://%s:%d/1?param_a=value_1&param_a=value_2&param_b=value_b", "localhost", envoyPort), &buf)
+			// put multi-value headers in the request
+			req.Header.Add("x-header-a", "value_1")
+			req.Header.Add("x-header-a", "value_2")
+			req.Header.Add("x-header-b", "value_b")
+
+			// execute request
+			res, err := http.DefaultClient.Do(req)
```
Comment (jbohanon) [inline test/e2e/aws_test.go:104]: nit: use a client with a timeout less than the Eventually loop

File: test/e2e/route_transformation_test.go [hunk 2]
```diff
@@ -416,4 +415,96 @@ var _ = Describe("Transformations", func() {
...
+			// log response body
+			body, err := ioutil.ReadAll(res.Body)
+			Expect(err).NotTo(HaveOccurred())
+			fmt.Println(string(body))
+
+			// parse the response body as JSON
+			var bodyJson map[string]interface{}
+			err = json.Unmarshal(body, &bodyJson)
+			Expect(err).NotTo(HaveOccurred())
+			bodyJson = bodyJson["json"].(map[string]interface{})
```
Comment (jbohanon) [inline test/e2e/route_transformation_test.go:467]: Why do we expect the body to be wrapped in this key `json`?

File: test/e2e/route_transformation_test.go [hunk 3]
```diff
@@ -416,4 +415,96 @@ var _ = Describe("Transformations", func() {
...
```
Comment (jbohanon) [inline test/e2e/route_transformation_test.go:499]: nit: This looks like it could be factored out

**Review by jbohanon — COMMENTED** [review]
Awesome!

**Comment by jbohanon** [comment]
We will be able to use envoy-gloo 1.25 after including the cluster extensions that were moved in upstream. This was resolved in envoy-gloo in [this commit](https://github.com/solo-io/envoy-gloo/pull/216/commits/9c0ef765bedae01b0e74f725b221deaa1d56d58d#diff-d98d375bf59fca540919a0e14e77ef9d4948eb8e45b82da7301377282fdf4a08) and merged with [this PR](https://github.com/solo-io/envoy-gloo/pull/216)

File: test/e2e/route_transformation_test.go (hunk 2 above)
Comment (ben-taussig-solo) [reply test/e2e/route_transformation_test.go:467]: This is a quirk of the httpbin /anything endpoint -- I'll leave a comment explaining this

File: test/e2e/aws_test.go [hunk 4]
```diff
@@ -93,7 +93,16 @@ var _ = Describe("AWS Lambda", func() {
 			var buf bytes.Buffer
 			buf.Write(body)
 
-			res, err := http.Post(fmt.Sprintf("http://%s:%d/1?param_a=value_1&param_b=value_b", "localhost", envoyPort), "application/octet-stream", &buf)
+			// form request with multi-value query params
+			req, err := http.NewRequest("POST", fmt.Sprintf("http://%s:%d/1?param_a=value_1&param_a=value_2&param_b=value_b", "localhost", envoyPort), &buf)
+			// put multi-value headers in the request
+			req.Header.Add("x-header-a", "value_1")
+			req.Header.Add("x-header-a", "value_2")
+			req.Header.Add("x-header-b", "value_b")
```
[the remark this answers is not in the document]
Comment (ben-taussig-solo) [reply test/e2e/aws_test.go]: Resolved here: https://github.com/solo-io/gloo/pull/7927/commits/ae4a0561db3ba1461dbdaede74bd69d4dae4b059

File: test/e2e/route_transformation_test.go [hunk 5]
```diff
@@ -416,4 +415,96 @@ var _ = Describe("Transformations", func() {
...
+			WriteVhost(vh)
+
+			// execute request -- expect a 200 response
+			url := fmt.Sprintf("http://%s:%d/anything?foo=bar&multiple=1&multiple=2", "localhost", envoyPort)
+			headers := map[string][]string{}
+			req := FormRequestWithUrlAndHeaders(url, headers)
+			res := GetSuccessfulResponse(req)
+
+			// log response body
+			body, err := ioutil.ReadAll(res.Body)
```
[the remark this answers is not in the document]
Comment (ben-taussig-solo) [reply test/e2e/route_transformation_test.go]: Removed ioutil from aws_test.go here: https://github.com/solo-io/gloo/pull/7927/commits/01c4559d9be899975e148174f4b6250405af9ad3

Resmoved ioutil from route_transformation_test.go here: https://github.com/solo-io/gloo/pull/7927/commits/66f03a038f6f63db3825ad617d06656c48ada023

File: test/e2e/route_transformation_test.go [hunk 6]
```diff
@@ -280,13 +281,11 @@ var _ = Describe("Transformations", func() {
 		}, "10s", "0.5s").ShouldNot(BeNil())
 	}
 
-	FormRequestWithUrlAndHeaders := func(url string, headers map[string]string) *http.Request {
+	FormRequestWithUrlAndHeaders := func(url string, headers map[string][]string) *http.Request {
```
[the remark this answers is not in the document]
Comment (ben-taussig-solo) [reply test/e2e/route_transformation_test.go]: I have unexported all helper methods from route_transformation_test.go here: https://github.com/solo-io/gloo/pull/7927/commits/dd94525a6d9ccb7b42f7439d74c9399d91bad121

None of these are referenced outside of the file they're defined in

File: test/e2e/route_transformation_test.go [hunk 7]
```diff
@@ -416,4 +415,96 @@ var _ = Describe("Transformations", func() {
...
+			Expect(err).NotTo(HaveOccurred())
+			fmt.Println(string(body))
+
+			// parse the response body as JSON
+			var bodyJson map[string]interface{}
+			err = json.Unmarshal(body, &bodyJson)
+			Expect(err).NotTo(HaveOccurred())
+			bodyJson = bodyJson["json"].(map[string]interface{})
+
+			// inspect the response body to confirm that the queryStringParameters were added to the metadata
```
[the remark this answers is not in the document]
Comment (ben-taussig-solo) [reply test/e2e/route_transformation_test.go:473]: Added a new comment clarifying this here: https://github.com/solo-io/gloo/pull/7927/commits/2a8d64d53153b487a2040190b973f516ec99b4f5

File: test/e2e/route_transformation_test.go [hunk 8]
```diff
@@ -416,4 +415,96 @@ var _ = Describe("Transformations", func() {
...
+			url := fmt.Sprintf("http://%s:%d/anything", "localhost", envoyPort)
+			headers := map[string][]string{
+				"x-solo-test-header": {"test"},
+				"foo":                {"bar", "baz"},
+			}
+			req := FormRequestWithUrlAndHeaders(url, headers)
+			res := GetSuccessfulResponse(req)
+
+			// log response body
+			body, err := ioutil.ReadAll(res.Body)
```
[the remark this answers is not in the document]
Comment (ben-taussig-solo) [reply test/e2e/route_transformation_test.go]: Resolved in the same commit as above

File: test/e2e/route_transformation_test.go [hunk 9]
```diff
@@ -416,4 +415,96 @@ var _ = Describe("Transformations", func() {
...
+			// execute request -- expect a 200 response
+			url := fmt.Sprintf("http://%s:%d/anything?foo=bar&multiple=1&multiple=2", "localhost", envoyPort)
+			headers := map[string][]string{}
+			req := FormRequestWithUrlAndHeaders(url, headers)
+			res := GetSuccessfulResponse(req)
+
+			// log response body
+			body, err := ioutil.ReadAll(res.Body)
+			Expect(err).NotTo(HaveOccurred())
+			fmt.Println(string(body))
```
[the remark this answers is not in the document]
Comment (ben-taussig-solo) [reply test/e2e/route_transformation_test.go]: I didn't intend to commit this; this was only here for debug purposes, so I opted to just remove the `println` statement

File: test/e2e/route_transformation_test.go (hunk 3 above)
Comment (ben-taussig-solo) [reply test/e2e/route_transformation_test.go]: Done here: https://github.com/solo-io/gloo/pull/7927/commits/bb30df8ff2680b7175144af82082a272f16ecd8a

File: docs/content/guides/traffic_management/request_processing/transformations/_index.md [hunk 10]
```diff
@@ -138,6 +138,7 @@ The `clearRouteCache` attribute is a boolean value that determines whether the r
 The `requestTransformation` and `responseTransformation` attributes have the {{< protobuf display="same format" name="transformation.options.gloo.solo.io.Transformation" >}} and specify transformations that will be applied to requests and responses respectively. The format can take one of two forms:
 
 - `headerBodyTransform`: this type of transformation will make all the headers available in the body. The result will be a JSON body that consists of two attributes: `headers`, containing the headers, and `body`, containing the original body.
+  - If `addRequestMetadata` is true, "queryString", "queryStringParameters", "multiValueQueryStringParameters", "httpMethod", "path", and "multiValueHeaders" will additionally be present in the body.
```
[the remark this answers is not in the document]
Comment (ben-taussig-solo) [reply docs/content/guides/traffic_management/request_processing/transformations/_index.md]: Resolved here: c642bf13baab63698c8bde3c61f65cfc8ea7e90e

**Comment by ben-taussig-solo** [comment]
> We will be able to use envoy-gloo 1.25 after including the cluster extensions that were moved in upstream. This was resolved in envoy-gloo in [this commit](https://github.com/solo-io/envoy-gloo/pull/216/commits/9c0ef765bedae01b0e74f725b221deaa1d56d58d#diff-d98d375bf59fca540919a0e14e77ef9d4948eb8e45b82da7301377282fdf4a08) and merged with [this PR](https://github.com/solo-io/envoy-gloo/pull/216)

Can confirm -- I ran these tests locally using a v1.25 build of envoy-gloo with the changes from my PR and they passed.

File: Makefile [hunk 11]
```diff
@@ -89,7 +89,9 @@ else
   endif
 endif
 
-ENVOY_GLOO_IMAGE ?= quay.io/solo-io/envoy-gloo:1.25.1-patch2
+# Note: We must update this to a released envoy-gloo image before merging this PR
```
Comment (sam-heilbron) [inline Makefile:92]: I think we can remove this comment, right?

File: test/e2e/aws_test.go [hunk 12]
```diff
@@ -63,8 +63,14 @@ var _ = Describe("AWS Lambda", func() {
 		envoyInstance *services.EnvoyInstance
 		secret        *gloov1.Secret
 		upstream      *gloov1.Upstream
+		httpClient    *http.Client
 	)
 
+	BeforeEach(func() {
+		httpClient = http.DefaultClient
+		httpClient.Timeout = 10 * time.Second
```
Comment (sam-heilbron) [inline test/e2e/aws_test.go:71]: Why do we need a timeout higher than the default?

File: test/e2e/route_transformation_test.go [hunk 13]
```diff
@@ -273,4 +275,131 @@ var _ = Describe("Transformations", func() {
...
+				WithRoutePrefixMatcher(e2e.DefaultRouteName, "/").
+				WithRouteActionToUpstream(e2e.DefaultRouteName, httpbinUpstream.Upstream).
+				Build()
+
+			testContext.ResourcesToCreate().Upstreams = gloov1.UpstreamList{httpbinUpstream.Upstream}
+			testContext.ResourcesToCreate().VirtualServices = v1.VirtualServiceList{vsToHtmlUpstream}
+		})
+
+		It("should handle queryStringParameters and multiValueQueryStringParameters", func() {
+			addRequestTransformation()
```
Comment (sam-heilbron) [inline test/e2e/route_transformation_test.go:363]: In each of these tests we create a VS in the BeforeEach, and then patch it in the test. Could we not just initialize the VS to include the transformation when we create it the first time? That way we don't need to write it two separate times (once to initialize, second to update)

File: test/e2e/route_transformation_test.go [hunk 14]
```diff
@@ -273,4 +275,131 @@ var _ = Describe("Transformations", func() {
...
+
+			// form request
+			req := formRequestWithUrlAndHeaders(fmt.Sprintf("http://localhost:%d/anything?foo=bar&multiple=1&multiple=2", defaults.HttpPort), nil)
+			// form matcher
+			matcher := &testmatchers.HttpResponse{
+				StatusCode: http.StatusOK,
+				Body: WithTransform(extractJson,
```
Comment (sam-heilbron) [inline test/e2e/route_transformation_test.go:370]: This is great! I think similar to how we added some multi-value header support (https://github.com/solo-io/gloo/blob/master/test/gomega/matchers/http.go#L65) in our custom utility, we should do a similar thing for our query param support. Do you think it would be possible to build a ContainsQueryString matcher that can be used in this test (and then other tests in the future)?

**Review by sam-heilbron — COMMENTED** [review]
Looking good! A few questions/thoughts that I'm happy to huddle/pair on further if you want

**Comment by ben-taussig-solo** [comment]
> Looking good! A few questions/thoughts that I'm happy to huddle/pair on further if you want

Thanks! Everything seems straightforward enough, I'll work through these on my own for a bit + reach out if I need clarification on anything

File: test/e2e/aws_test.go (hunk 12 above)
Comment (ben-taussig-solo) [reply test/e2e/aws_test.go:71]: DefaultClient has no default timeout. I think I saw this article in slack a while ago: https://medium.com/@nate510/don-t-use-go-s-default-http-client-4804cb19f779

> TL;DR: Go’s http package doesn’t specify request timeouts by default, allowing services to hijack your goroutines. Always specify a custom http.Client when connecting to outside services.

I added some log statements around line 71 and confirmed that the initial timeout is unset
```
Logging initial DefaultClient timeout:  0s
Logging updated DefaultClient timeout:  10s
```

Comment (sam-heilbron) [reply test/e2e/aws_test.go:71]: Nice! Is this something specific to the lambda tests, or something we could institute more broadly. I really like it and almost wonder if we could just build a simple client with a set timeout for all tests to use. Certainly not required for this work, thanks for the explanation

File: Makefile (hunk 11 above)
Comment (ben-taussig-solo) [reply Makefile]: Removed here: https://github.com/solo-io/gloo/pull/7927/commits/3fd9ab3a835ff70f5b284dbe034c86cf02c6066b

File: test/e2e/route_transformation_test.go (hunk 13 above)
Comment (ben-taussig-solo) [reply test/e2e/route_transformation_test.go]: made that change here: https://github.com/solo-io/gloo/pull/7927/commits/a224e8c627c18288e46fa2d04f2581d8a686ec43

File: test/e2e/route_transformation_test.go (hunk 14 above)
Comment (ben-taussig-solo) [reply test/e2e/route_transformation_test.go:362]: I think this might not be relevant here -- We're trying to match a json response body that contains fields like the following instead of the unprocessed querystring

```json
{
  "queryStringParameters": {
    "foo": "bar",
    "multiple": "2",
  },
  "multiValueQueryStringParameters": {
    "multiple": [
      "2",
      "3"
    ]
  }
}
```

File: test/e2e/route_transformation_test.go [hunk 15]
```diff
@@ -273,4 +275,131 @@ var _ = Describe("Transformations", func() {
...
+			if bodyJson["json"] == nil {
+				return nil
+			}
+
+			return bodyJson["json"].(map[string]interface{})
+		}
+
+		BeforeEach(func() {
+			// create an upstream for the httpbin service
+			httpbinUpstream := v1helpers.NewTestUpstream("httpbin.org", []uint32{80}, nil)
```
[the remark this answers is not in the document]
Comment (ben-taussig-solo) [reply test/e2e/route_transformation_test.go]: Made that change here: https://github.com/solo-io/gloo/pull/7927/commits/88aa4a2767ba4801d976e6cdef53487ea1085b0d

File: test/e2e/route_transformation_test.go [hunk 16]
```diff
@@ -273,4 +275,115 @@ var _ = Describe("Transformations", func() {
 			EventuallyHtmlResponseTransformed().Should(Succeed())
 		})
 	})
+
+	Context("requestTransformation", func() {
+		// form a request with the given headers
+		// note that the Host header is set to the default host
+		formRequestWithUrlAndHeaders := func(url string, headers map[string][]string) *http.Request {
+			// form request
+			req, err := http.NewRequest("GET", url, nil)
```
Comment (sam-heilbron) [inline test/e2e/route_transformation_test.go:284]: Let's use the constant provided: http.MethodGet

File: test/e2e/route_transformation_test.go [hunk 17]
```diff
@@ -273,4 +275,115 @@ var _ = Describe("Transformations", func() {
...
+			return req
+		}
+
+		// send the given request and assert that the response matches the given expected response
+		eventuallyRequestMatches := func(req *http.Request, expectedResponse *testmatchers.HttpResponse) AsyncAssertion {
+			return Eventually(func(g Gomega) {
+				// send request
+				client := &http.Client{Timeout: 10 * time.Second}
+				res, err := client.Do(req)
+				Expect(err).NotTo(HaveOccurred())
```
Comment (sam-heilbron) [inline test/e2e/route_transformation_test.go:297]: let's use g.Expect

File: test/e2e/route_transformation_test.go [hunk 18]
```diff
@@ -273,4 +275,115 @@ var _ = Describe("Transformations", func() {
...
+
+			// create a virtual host with a route to the upstream
+			vsToEchoUpstream := helpers.NewVirtualServiceBuilder().
+				WithName(e2e.DefaultVirtualServiceName).
+				WithNamespace(writeNamespace).
+				WithDomain(e2e.DefaultHost).
+				WithRoutePrefixMatcher(e2e.DefaultRouteName, "/").
+				WithRouteActionToUpstream(e2e.DefaultRouteName, echoUpstream.Upstream).
+				Build()
+
```
Comment (sam-heilbron) [inline test/e2e/route_transformation_test.go:327]: We expose WithVirtualHostOptions https://github.com/solo-io/gloo/blob/master/test/helpers/virtual_services.go#L67 in the builder. Can we use that, instead of building and then updating? The idea of the builder is to make it easier to apply piecewise configuration

File: test/e2e/route_transformation_test.go [hunk 19]
```diff
@@ -273,4 +275,115 @@ var _ = Describe("Transformations", func() {
...
+			if err != nil {
+				return nil
+			}
+
+			return bodyJson
+		}
+
+		BeforeEach(func() {
+			// create an upstream for the httpbin service
+			echoUpstream := v1helpers.NewTestHttpUpstreamWithReply(testContext.Ctx(), testContext.EnvoyInstance().LocalAddr(), "")
```
Comment (sam-heilbron) [inline test/e2e/route_transformation_test.go:317]: We shouldn't need this. The TestContext gives us an upstream for free (https://github.com/solo-io/gloo/blob/master/test/e2e/test_context.go#L80) because so all tests use one

File: test/e2e/route_transformation_test.go [hunk 20]
```diff
@@ -273,4 +275,115 @@ var _ = Describe("Transformations", func() {
...
+										},
+									},
+								},
+							},
+						},
+					},
+				},
+			}
+
+			testContext.ResourcesToCreate().Upstreams = gloov1.UpstreamList{echoUpstream.Upstream}
```
Comment (sam-heilbron) [inline test/e2e/route_transformation_test.go:346]: Similar to above, we shouldn't need to update the upstreams because the testContext gives us what we need

File: test/e2e/route_transformation_test.go [hunk 21]
```diff
@@ -273,4 +275,115 @@ var _ = Describe("Transformations", func() {
...
+				client := &http.Client{Timeout: 10 * time.Second}
+				res, err := client.Do(req)
+				Expect(err).NotTo(HaveOccurred())
+
+				g.Expect(res).To(testmatchers.HaveHttpResponse(expectedResponse))
+			}, "10s", ".5s")
+		}
+
+		// used to extract the JSON response body from the echo upstream
+		extractJson := func(b []byte) map[string]interface{} {
```
Comment (sam-heilbron) [inline test/e2e/route_transformation_test.go:304]: I think this could be a shared transform https://github.com/solo-io/gloo/blob/master/test/gomega/transforms/transforms.go. It may be simple, but I think it's powerful to have those defined in a single place for others to easily use

File: test/v1helpers/test_upstream.go [hunk 22]
```diff
@@ -116,17 +116,17 @@ func NewTestGRPCUpstream(ctx context.Context, addr string, replicas int) *TestUp
...
 	return us
 }
 
 var testUpstreamId = 0
 
-// newTestUpstream creates a static Upstream that can route traffic to a set of ports for a given address
+// NewTestUpstream creates a static Upstream that can route traffic to a set of ports for a given address
 // It contains a unique name (since tests may run in parallel), with a suffix id that increases each invocation
-func newTestUpstream(addr string, ports []uint32, responses <-chan *ReceivedRequest) *TestUpstream {
+func NewTestUpstream(addr string, ports []uint32, responses <-chan *ReceivedRequest) *TestUpstream {
```
Comment (sam-heilbron) [inline test/v1helpers/test_upstream.go:129]: Why do we export this method? The intention is that this is private, and all the "NewTestUpstreamWith...etc" are exported. I've also found this file to be pretty confusing, so the TestContext creates a TestUpstream for you

File: test/e2e/route_transformation_test.go (hunk 18 above)
Comment (ben-taussig-solo) [reply test/e2e/route_transformation_test.go:327]: oh nice, I missed that one

File: test/v1helpers/test_upstream.go (hunk 22 above)
Comment (ben-taussig-solo) [reply test/v1helpers/test_upstream.go]: Resolved here: https://github.com/solo-io/gloo/pull/7927/commits/168f8be935859f885fd92a0b58c1360420bd1146

File: test/e2e/route_transformation_test.go (hunk 21 above)
Comment (ben-taussig-solo) [reply test/e2e/route_transformation_test.go]: Resolved here: https://github.com/solo-io/gloo/pull/7927/commits/168f8be935859f885fd92a0b58c1360420bd1146

File: test/e2e/route_transformation_test.go (hunk 20 above)
Comment (ben-taussig-solo) [reply test/e2e/route_transformation_test.go]: Resolved here: https://github.com/solo-io/gloo/pull/7927/commits/168f8be935859f885fd92a0b58c1360420bd1146

File: test/e2e/route_transformation_test.go (hunk 19 above)
Comment (ben-taussig-solo) [reply test/e2e/route_transformation_test.go]: Resolved here: https://github.com/solo-io/gloo/pull/7927/commits/168f8be935859f885fd92a0b58c1360420bd1146

File: test/e2e/route_transformation_test.go (hunk 17 above)
Comment (ben-taussig-solo) [reply test/e2e/route_transformation_test.go]: Resolved here: https://github.com/solo-io/gloo/pull/7927/commits/168f8be935859f885fd92a0b58c1360420bd1146

File: test/e2e/route_transformation_test.go (hunk 16 above)
Comment (ben-taussig-solo) [reply test/e2e/route_transformation_test.go]: Resolved here: https://github.com/solo-io/gloo/pull/7927/commits/168f8be935859f885fd92a0b58c1360420bd1146

File: changelog/v1.14.0-beta17/request-transform-improvements.yaml [hunk 23]
```diff
@@ -0,0 +1,11 @@
...
+- type: DEPENDENCY_BUMP
+  dependencyOwner: solo-io
+  dependencyRepo: envoy-gloo
+  dependencyTag: v1.25.1-patch3
+  description: >
+   Update envoy-gloo to v1.25.1-patch3, pulling in changes that add 
+   queryStringParameters, multiValueQueryStringParameters, and
+   multiValueHeaders fields to headerBodyTransform output when
+   `add_request_metadata` is enabled.
+  issueLink: https://github.com/solo-io/solo-projects/issues/4571
```
Comment (sam-heilbron) [inline changelog/v1.14.0-beta17/request-transform-improvements.yaml:11]: nit: I know you didn't write this issue. In the future I think most issues should be in the Gloo repo, especially since this functionality is open source. It allows users to easily track the progress of the issue and any linked PRs. I think in the future we can create duplicates if necessary, or push product to open issues in there

File: test/e2e/route_transformation_test.go [hunk 24]
```diff
@@ -273,4 +276,98 @@ var _ = Describe("Transformations", func() {
...
+							},
+						},
+					},
+				}).
+				Build()
+
+			testContext.ResourcesToCreate().VirtualServices = v1.VirtualServiceList{vsToEchoUpstream}
+		})
+
+		It("should handle queryStringParameters and multiValueQueryStringParameters", func() {
```
Comment (sam-heilbron) [inline test/e2e/route_transformation_test.go:334]: These are great tests with really expressive assertions

**Review by sam-heilbron — APPROVED** [review]
Great work! I think this nicely documents the new behavior, introduces tests to demonstrate the functionality, and expands our internal testing libraries.

**Comment by ashishb-solo** [comment]
Looks good to me overall and I'm happy to approve. Just one thing - have you tested this where `multiValueQueryStringParameters` has 3 or 4 elements? Given the nature of the logic required for creating this field, I think it would just be good to make sure that it handles odd/even values >2 properly. My concern is that the nature of the lambda requirements make it possible to accidentally add duplicate fields or omit ones.

Arguably, this is more of an issue in the data plane than the control plane, but it's easier to write tests here, so I'd be happy just to see it here.

**Comment by ben-taussig-solo** [comment]
> Looks good to me overall and I'm happy to approve. Just one thing - have you tested this where `multiValueQueryStringParameters` has 3 or 4 elements? Given the nature of the logic required for creating this field, I think it would just be good to make sure that it handles odd/even values >2 properly. My concern is that the nature of the lambda requirements make it possible to accidentally add duplicate fields or omit ones.
> 
> Arguably, this is more of an issue in the data plane than the control plane, but it's easier to write tests here, so I'd be happy just to see it here.

Added tests that validate the 3 and 4 element cases here: https://github.com/solo-io/gloo/pull/7927/commits/4d16df1184eab4a797df92d50b4426fde61dfa98 . Let me know what you think!

**Review by ashishb-solo — APPROVED** [review]

**Review by sam-heilbron — APPROVED** [review]
